design_documents/test_design.md

```diff
@@ -0,0 +1,30 @@
+# Test design
+
+## Context
+
+Current test design in Starknet Forge has some issues:
+- gas estimation is problematic since we have to distinguish between pure Cairo gas
+(e.g., creating an array in test) and Starknet gas
+- starknet functions like `get_caller_address` are unavailable in tests directly
+- starknet system calls like `library_call` are unavailable in tests directly
+- testing internal contract functions requires exposing them as `external`
+- operating on the contract execution layer while tests being pure Cairo code can create confusion
+
+## Goal
+
+Propose a solution of test design that solves most of the aforementioned issues.
+
+## Proposed Solution
+
+Tests should be contracts (just like in the early days of Protostar).
+Basically, the code in test should behave as if it was a function called in an invoke transaction.
```

Comment:
Should we still be allowing the patter of `declare & deploy` and then using dispatchers if this change is introduced?  If so, should the contract-test state be applied to this calls? E.g. should the caller address be that of whatever it was set in the test itself (if we allow it)? 

code example
```cairo
#[test]
fn test() {
  cheatcode_that_changes_the_caller_address_in_test(1234);
  assert(get_caller_address() == 1234, 'this passes');
  // ...
  other_dispatcher.function();
}

// ...
#[starknet::contract]
mod Other {
  // ...
  fn function() {
    // should get_caller_address here be whatever it was set in the test above i.e. 1234?
    get_caller_address() == ???
  }
}

Reply:
Discussed offline - all tests using the declare & deploy flow will still work, these functions will basically be syscalls with better UX.
In your example `get_caller_address` inside `function` should return the address of the test (since each test is a contract), so chaging caller address inside the test would only change the address of the artificial "caller of the `test` function". We can think of a test as a function on starknet - basically everything happens on execution layer (or inside `invoke` when looking from 3rd layer perspective)